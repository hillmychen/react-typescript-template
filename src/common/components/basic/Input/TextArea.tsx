import React from 'react'
import { Input as AntdInput } from 'antd'
import { TextAreaProps as AntdTextAreaProps } from 'antd/lib/input'
import classnames from 'classnames'
import styles from './index.module.less'

export interface TextAreaProps extends AntdTextAreaProps {
  /** 禁止输入空格 */
  trim?: boolean
  /** 使 Textarea 透明 */
  transparent?: boolean
}

export class TextArea extends React.PureComponent<TextAreaProps> {
  private readonly inputRef: React.RefObject<any> = React.createRef()

  private onChange = (event: React.ChangeEvent<HTMLTextAreaElement>) => {
    const { trim, onChange } = this.props
    if (trim) {
      event.target.value = event.target.value.trim()
    }
    onChange?.(event)
  }

  focus = () => this.inputRef.current?.focus()

  blur = () => this.inputRef.current?.blur()

  render() {
    const { onChange, trim, className, transparent, ...restTextAreaProps } = this.props
    return <AntdInput.TextArea ref={this.inputRef} {...restTextAreaProps} className={classnames(className, { [styles.transparent]: transparent })} onChange={this.onChange} />
  }
}

export default TextArea
