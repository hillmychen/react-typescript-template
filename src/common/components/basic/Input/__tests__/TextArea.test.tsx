import React from 'react'
import '@testing-library/jest-dom/extend-expect'
import { render, fireEvent } from '@testing-library/react'
import { TextArea } from '../TextArea'

test('TextArea onChange Event', () => {
  const { asFragment, getByRole } = render(<TextArea />)
  const input = getByRole('textbox', { hidden: true }) as HTMLTextAreaElement
  fireEvent.change(input, { target: { value: 'test value' } })
  expect(input.value).toEqual('test value')
  expect(asFragment()).toMatchSnapshot()
})

test('TextArea test trim', done => {
  const { asFragment, getByRole } = render(<TextArea trim onChange={handlerOnChange} />)
  const input = getByRole('textbox', { hidden: true }) as HTMLTextAreaElement
  fireEvent.change(input, { target: { value: ' test value ' } })
  function handlerOnChange(event: React.ChangeEvent<HTMLTextAreaElement>) {
    expect(event.target.value).toEqual('test value')
    expect(asFragment()).toMatchSnapshot()
    done()
  }
})
