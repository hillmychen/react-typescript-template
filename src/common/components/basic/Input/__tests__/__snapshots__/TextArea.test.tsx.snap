// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextArea onChange Event 1`] = `
<DocumentFragment>
  <textarea
    class="ant-input"
  >
    test value
  </textarea>
</DocumentFragment>
`;

exports[`TextArea test trim 1`] = `
<DocumentFragment>
  <textarea
    class="ant-input"
  />
</DocumentFragment>
`;
